src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,202 @@
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(getBoundariesMessage(index));
+        }
+    }
+
+    private void unlinkNode(Node<T> nodeToRemove) {
+        if (nodeToRemove.lastNode == null) {
+            unlinkHead();
+        } else if (nodeToRemove.nextNode == null) {
+            unlinkTail();
+        } else {
+            nodeToRemove.lastNode.nextNode = nodeToRemove.nextNode;
+            nodeToRemove.nextNode.lastNode = nodeToRemove.lastNode;
+            size--;
+        }
+    }
+
+    private void unlinkTail() {
+        Node<T> newTail = tail.lastNode;
+        tail.nodeValue = null;
+        tail.lastNode = null;
+        tail = newTail;
+        if (newTail == null) {
+            head = null;
+        } else {
+            tail.nextNode = null;
+        }
+        size--;
+    }
+
+    private void unlinkHead() {
+        Node<T> newHead = head.nextNode;
+        head.nodeValue = null;
+        head.nextNode = null;
+        head = newHead;
+        if (newHead == null) {
+            tail = null;
+        } else {
+            head.lastNode = null;
+        }
+        size--;
+    }
+
+    private void addTail(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        Node<T> tempTail = tail;
+        tail = newNode;
+        if (tempTail == null) {
+            head = newNode;
+        } else {
+            tempTail.nextNode = newNode;
+        }
+        size++;
+    }
+
+    private void addBetween(T value, Node<T> indexNode) {
+        Node<T> newNode = new Node<>(indexNode.lastNode, value, indexNode);
+        indexNode.lastNode = newNode;
+        if (newNode.lastNode == null) {
+            head = newNode;
+        } else {
+            newNode.lastNode.nextNode = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> nodeByIndex(int index) {
+        if (index <= size / 2) {
+            int counter = 0;
+            Node<T> currentNode = head;
+            while (currentNode != null) {
+                if (counter == index) {
+                    return currentNode;
+                }
+                currentNode = currentNode.nextNode;
+                counter++;
+            }
+        } else {
+            int counter = size - 1;
+            Node<T> currentNode = tail;
+            while (currentNode != null) {
+                if (counter == index) {
+                    return currentNode;
+                }
+                currentNode = currentNode.lastNode;
+                counter--;
+            }
+        }
+        return null;
+    }
+
+    private Node<T> findFirstNodeByValue(T value) {
+        Node<T> currenNode = head;
+        while (currenNode != null) {
+            if (currenNode.nodeValue == value
+                    || (currenNode.nodeValue != null && currenNode.nodeValue.equals(value))) {
+                return currenNode;
+            }
+            currenNode = currenNode.nextNode;
+        }
```

Comment:
method is used only once, better move this logic to remove by value. Throwing exception here and catching in remove isn't the best way to go